### ISSUE TITLE
Ability to ignore certain productions

### ISSUE THREAD

**Issue by chromaticbum** [pegjs/pegjs#11]
It would be nice to be able to tell the lexer/parser to ignore certain productions (i.e. whitespace and comment productions) so that it becomes unnecessary to litter all other productions with comment/whitespace allowances. This may not be possible though, due to the fact that lexing is builtin with parsing?

Thank you

**Comment by StreetStrider** [comment]
Hello, I've created [a snippet](https://gist.github.com/StreetStrider/acf2b870b751e6267487) to ignore some general cases: `null`, `undefined` and strings with only space symbols.
It can be required in the head of the grammar file, like:
```js
{
  var strip = require('./strip-ast');
}
```
The two ways to improve it:
* Customizable filter for terms — to ignore the specific terms that certain grammar is required.
* Skip nested empty arrays — this can be done on second stage after `strip`, it'll remove «pyramids» of nested empty arrays.
If anyone interested, we can upgrade it to a package.

**Comment by atesgoral** [comment]
@richb-hanover Where did your ASN.1 definition parser efforts land?

**Comment by richb-hanover** [comment]
@atesgoral - I bailed out. I didn't need a "real parser" - I only needed to isolate certain named elements in the target file. 

So I did what any wimpy guy would do - used regular expressions. (And then I had two problems :-)

But it did the trick, so I was able to move on to the next challenge. Good luck in your project!

**Comment by Izhaki** [comment]
Having had a look at [chevrotain and its skip option](http://sap.github.io/chevrotain/docs/tutorial/step1_lexing.html#skipping-tokens), something like this is hugely desirable.

Too often we find ourselves writing something like this:

```
Pattern = head:PatternPart tail:( WS "," WS PatternPart )*
{
  return {
    type: 'pattern',
    elements: buildList( head, tail, 3 )
  };
}
```
Would be cool if we could write this instead:

```
WS "whitespace" = [ \t\n\r] { return '@@skipped' }

IgnoredComma = "," { return '@@skipped' }

Pattern = head:PatternPart tail:( WS IgnoredComma WS PatternPart )*
{
  return {
    type: 'pattern',
    elements: [head].concat(tail)
  };
}
```

**Comment by atesgoral** [comment]
@richb-hanover, and anybody else who got here in search for a similar need, I ended up writing my own parsers, too: https://www.npmjs.com/package/asn1exp and https://www.npmjs.com/package/asn1-tree

**Comment by StoneCypher** [comment]
A skip would be relatively easy to implement using `es6 symbol`, or maybe more durably by passing the parser a predicate at parse time (I prefer the latter option)

**Comment by darlanalves** [comment]
Just stumbled upon this too.
Not knowing anything about the innards of PEG.js, lemme throw a bone out there...

When we write a rule, at the end of it we can add a return block. 
In that block, we can call things like `text()` and `location()`. These are internal functions.

Somewhere in the code the returned value of that block goes into the output stream.

So what would need to change in PEG.js if I want to skip a value returned by a rule if that value is the return of callking a `skip` local function ?

e.g. `comment = "//" space ([^\n])* newline { return skip() }`

As mentioned above, skip() could return a Symbol, which is then checked by the code somewhere and removed.
Something like what lzhaki said, but internal to the library

**Comment by Mingun** [comment]
I don't understand your question. Are you looking for a way to fail a rule under some circumstances? Use `&{...}` or `!{...}`. Otherwise just don't use returned value of `comment` rule:
```
seq = comment r:another_rule { return r; };
choice = (comment / another_rule) { <you need to decide what to return instead of "comment" result> };
```

**Comment by stoneRdev** [comment]
If it helps anyone, I ignore whitespace by having my top-level rule filter the array of results.

Example: 
```

program
	= prog:expression+ {return prog.filter(a => a)}

expression
	= float
    / number
    / whitespace

float
	= digits:(number"."number) {return parseFloat(digits.join(""),10)}
    
number 
	= digits:digit+ {return parseInt(digits.join(""),10)}
    
digit 
	= [0-9]
    
whitespace
	= [ \t\r\n] {return undefined}
```

This will happily parse input while keeping whitespace out of the result array.
This will also work for things like comments, just have the rule return undefined and the top level rule will filter it out

**Comment by StoneCypher** [comment]
That only works for top-level productions.  You have to manually filter every parent that could contain a filterable child.

**Comment by stoneRdev** [comment]
@StoneCypher True, it does require some top level work, but it works for me, and I think as long as the gammar isn't too complex, one should be able to get away with having a top level filter. 

Other than that, all I can think of is have a top level function that filters whitespace from input and pass every match through it. Slower for sure, and requires alot more calls, but easy if you (like me) pass everything into a token generator. You can call the filter function from where you generate tokens, and only have to worry about generating your tokens and the whitespace is more or less automatically filtered

**Comment by hildjj** [comment]
One of the things I liked about the current HEAD of pegjs is its (undocumented) support for picking fields without having to create labels and do return statements.  It looks like this:

```pegjs
foo = @bar _ @baz
bar = $"bar"i
baz = $"baz"i
_ = " "*
```
```js
parse('barbaz') // returns [ 'bar', 'baz' ]
```

I feel like this gives nice, clean, explicit syntax for this use case plus a bunch of others.

**Comment by markus-fe** [comment]
@hildjj This is exactly what I needed in combination with [parsing lists](https://peggyjs.org/documentation#parsing-lists). Peggy is wonderful, thank you for your effort!

I guess, this example also would make a good candidate for the documentation page, as it illustrates the usage of `@` (which I didn't quite understand from reading the docs).
